File: day15/puzzle.js
const fs = require("fs");
const readLines = (path) => {
  return fs
    .readFileSync(path, { encoding: "utf8" })
    .split("\n")
    .filter(Boolean)
    .map((line) => line.match(/-?\d+/g).map(Number));
}

const sensorMap = readLines(process.argv[2]).reduce((map, line) => {
  const [sx, sy, bx, by] = line;
  const manhattanDist = Math.abs(sx - bx) + Math.abs(sy - by);
  map.set([sx, sy], manhattanDist);
  return map;
}, new Map());

const merge = (intervals) => {
  intervals.sort((a, b) => a[0] - b[0]);

  return intervals.reduce((merged, interval) => {
    const [first, last] = interval
    if (merged.length === 0 || merged.at(-1)[1] < first) {
      merged.push([first, last]);
    } else {
      merged.at(-1)[1] = Math.max(merged.at(-1)[1], last);
    }
    return merged;
  }, [])
}

const impossibleIntervalFor = (row) => {
  const intervals = [];

  for (const [[x, y], manDist] of sensorMap) {
    const vertDist = Math.abs(y - row);
    if (vertDist >= manDist) continue;

    const horiDist = manDist - vertDist;
    const interval = [x - horiDist, x + horiDist];
    intervals.push(interval);
  }
  return merge(intervals);
}

const countImpossibleBeaconPositionsFor = (row) => {
  const [left, right] = impossibleIntervalFor(row).flat()
  return Math.abs(left - right);
}

console.log(countImpossibleBeaconPositionsFor(2_000_000)); // Part1

const findDistressBeacon = (searchSize) => {
  for (let row = 0; row < searchSize; row++) {
    const rowIntervals = impossibleIntervalFor(row);

    if (rowIntervals.length > 1) {
      const distressBeaconX = rowIntervals[0][1];
      return (distressBeaconX * 4_000_000) + row;
    }
  }
}

console.log(findDistressBeacon(4_000_000)); // Part2

File: day15/puzzle.rb
# frozen_string_literal: true

input = File.readlines(*ARGV, chomp: true)
            .map { |l| l.scan(/\d+|-\d+/).map(&:to_i) }

@sensor_map = input.each_with_object({}) do |(sx, sy, bx, by), map|
  manhattan_dist = (sx - bx).abs + (sy - by).abs
  map[[sx, sy]] = manhattan_dist
end

def merge(intervals)
  intervals.sort_by!(&:first)

  intervals.each_with_object([]) do |(first, last), merged|
    if merged.empty? || merged[-1][1] < first
      merged << [first, last]
    else
      merged[-1][1] = [merged[-1][1], last].max
    end
  end
end

def impossible_beacon_interval_for(row)
  intervals = []

  @sensor_map.each do |sensor, man_dist|
    x, y = sensor
    vert_dist = (y - row).abs

    next if vert_dist >= man_dist

    hori_dist = man_dist - vert_dist
    interval = [x - hori_dist, x + hori_dist]
    intervals << interval
  end
  merge(intervals)
end

def count_impossible_beacon_positions_for(row)
  left, right = impossible_beacon_interval_for(row).flatten
  (left - right).abs
end

p count_impossible_beacon_positions_for(2_000_000) # p1-> 4724228

def find_distress_beacon(search_space)
  search_space.times do |row|
    row_intervals = impossible_beacon_interval_for(row)

    if row_intervals.length > 1
      distress_beacon_x = row_intervals[0][1]
      return (distress_beacon_x * 4_000_000) + row
    end
  end
end

p find_distress_beacon(4_000_000) # p2-> 13622251246513
